engines/rados.c

```diff
@@ -0,0 +1,501 @@
+{
+	struct rados_data *rados = td->io_ops_data;
+	struct fio_rados_iou *fri = io_u->engine_data;
+	char* object = rados->objects[io_u->file->engine_pos];
+	int r = -1;
+
+	fio_ro_check(td, io_u);
+
+	fri->io_seen = 0;
+	fri->io_complete = 0;
+
+	if (io_u->ddir == DDIR_WRITE) {
+		r = rados_aio_create_completion(io_u,
+		on_complete_cb,
+		NULL /*cb_safe*/,
+		&fri->completion);
+		if (r < 0) {
+			log_err("rados_aio_create_completion failed.\n");
+			goto failed;
+		}
+
+		r = rados_aio_write(
+		rados->io_ctx,
+		object,
+		fri->completion,
+		io_u->xfer_buf,
+		io_u->xfer_buflen,
+		io_u->offset);
+		if (r < 0) {
+			log_err("rados_write failed.\n");
+			goto failed_comp;
+		}
+		return FIO_Q_QUEUED;
+	} else if (io_u->ddir == DDIR_READ) {
+		r = rados_aio_create_completion(io_u,
+			on_complete_cb,
+			NULL /*cb_safe*/,
+			&fri->completion);
+		if (r < 0) {
+			log_err("rados_aio_create_completion failed.\n");
+			goto failed;
+		}
+		r = rados_aio_read(rados->io_ctx,
+			object,
+			fri->completion,
+			io_u->xfer_buf,
+			io_u->xfer_buflen,
+			io_u->offset);
+		if (r < 0) {
+			log_err("rados_aio_read failed.\n");
+			goto failed_comp;
+		}
+		return FIO_Q_QUEUED;
+	} else if (io_u->ddir == DDIR_TRIM) {
+		r = rados_aio_create_completion(io_u,
+			on_write_op_complete_cb,
+			NULL /*cb_safe*/,
+			&fri->completion);
+		if (r < 0) {
+			log_err("rados_aio_create_completion failed.\n");
+			goto failed;
+		}
+		fri->write_op = rados_create_write_op();
+		if (fri->write_op == NULL) {
+			log_err("rados_create_write_op failed.\n");
+			goto failed_comp;
+		}
+		rados_write_op_zero(
+			fri->write_op,
+			io_u->offset,
+			io_u->xfer_buflen);
+		r = rados_aio_write_op_operate(
+			fri->write_op,
+			rados->io_ctx,
+			fri->completion,
+			object,
+			NULL, 
+			0);
+
+		if (r < 0) {
+			log_err("rados_aio_write_op_operate failed.\n");
+			goto failed_write_op;
+		}
+		return FIO_Q_QUEUED;
+	 }
+
+	log_err("WARNING: Only DDIR_READ, DDIR_WRITE and DDIR_TRIM are supported!");
+
+failed_write_op:
+	rados_release_write_op(fri->write_op);
+failed_comp:
+	rados_aio_release(fri->completion);
+failed:
+	io_u->error = -r;
+	td_verror(td, io_u->error, "xfer");
+	return FIO_Q_COMPLETED;
+}
+
+static struct io_u *fio_rados_event(struct thread_data *td, int event)
+{
+	struct rados_data *rados = td->io_ops_data;
+	return rados->aio_events[event];
+}
+
+int fio_rados_getevents(struct thread_data* td, unsigned int min,
+  unsigned int max, const struct timespec* t)
+{
+	struct rados_data *rados = td->io_ops_data;
+	unsigned int events = 0;
+	struct io_u* u;
+	struct fio_rados_iou *fri;
+	unsigned int i;
+
+	// loop through inflight ios until we find 'min' completions
+	do {
+		io_u_qiter(&td->io_u_all, u, i) {
+			if (!(u->flags & IO_U_F_FLIGHT))
+				continue;
+
+			fri = u->engine_data;
+			if (fri->io_complete) {
+				fri->io_complete = 0;
+				rados->aio_events[events] = u;
+				events++;
+			}
+		}
+		if (events >= min)
+			break;
+		usleep(100);
```

Comment:
You should not sleep unless ```t``` is set and indicates that the callers wants to wait for completions.